src/main/java/io/rocketpartners/cloud/utils/SqlUtils.java

```diff
@@ -431,6 +431,60 @@ public static String buildInsertSQL(Connection conn, String tableName, Object[]
       return sql.toString();
    }
 
+   public static String buildInsertOnDuplicateKeySQL(Connection conn, String tableName, Object[] columnNameArray)
+   {
+      StringBuffer sql = new StringBuffer(buildInsertSQL(conn, tableName, columnNameArray));
+      sql.append(" ON DUPLICATE KEY UPDATE ");
+      for (Object col : columnNameArray)
+      {
+         sql.append(col).append("=VALUES(").append(col).append("), ");
+      }
+      sql.substring(0, sql.length() - 1);
+      return sql.toString();
+   }
+
+   public static String mysqlUpsert(Connection conn, String tableName, List<Map<String, Object>> rows) throws Exception
+   {
+      LinkedHashSet keySet = new LinkedHashSet();
+
+      for (Map row : rows)
+      {
+         keySet.addAll(row.keySet());
+      }
+
+      List<String> keys = new ArrayList(keySet);
+      String sql = buildInsertOnDuplicateKeySQL(conn, tableName, keys.toArray());
+
+      Exception ex = null;
+      PreparedStatement stmt = conn.prepareStatement(sql);
+      try
+      {
+         notifyBefore("mysqlUpsert", sql, rows);
+
+         for (Map<String, Object> row : rows)
+         {
+            for (int i = 0; i < keys.size(); i++)
+            {
+               Object value = row.get(keys.get(i));
+               ((PreparedStatement) stmt).setObject(i + 1, value);
+            }
+            stmt.addBatch();
+         }
+         stmt.executeBatch();
+      }
+      catch (Exception e)
+      {
+         ex = e;
+         notifyError("mysqlUpsert", sql, rows, ex);
+      }
+      finally
+      {
+         SqlUtils.close(stmt);
+         notifyAfter("mysqlUpsert", sql, rows, ex, null);
+      }
+      return sql;
```

Comment:
We don't want to return the sql....we want to return a list of the keys that were generated by the upsert.